### PR TITLE
added workflow for cancelling a line item

### PR DESCRIPTION
**Pull request by shreyas-satish**

### CODE REVIEW

File: boxoffice/models/line_item.py [hunk 1]
```diff
@@ -109,6 +109,17 @@ def confirm(self):
...
+    @property
+    def is_confirmed(self):
+        return self.status == LINE_ITEM_STATUS.CONFIRMED
+
+    def cancel(self):
+        """
+        Sets status and cancelled_at.
+        """
+        self.status = LINE_ITEM_STATUS.CANCELLED
+        self.cancelled_at = datetime.datetime.utcnow()
```
Comment (jace) [inline boxoffice/models/line_item.py]: Use `func.utcnow()`

File: boxoffice/templates/line_item_cancellation_mail.html [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+</head>
+
+<body>
+  <div>
+
+    <p>Hello {{ order.buyer_fullname }},</p>
+
+    <p>With reference to your order (Receipt No: {{ order.invoice_no }}), "{{ item_title }}" has been cancelled.</p>
+    {%- if is_paid %}
+      <p>A refund of &#8377; {{line_item.final_amount}} has been initiated. You should receive the refund in 2-5 days.</p>
```
Comment (jace) [inline boxoffice/templates/line_item_cancellation_mail.html]: The rupee symbol should ideally come from code passing in currency symbol and value.

File: boxoffice/views/order.py [hunk 3]
```diff
@@ -314,6 +315,35 @@ def format_assignee(line_item):
...
 
+# TODO activate route when front-end becomes available
+# @app.route('/line_item/<line_item_id>/cancel', methods=['POST'])
+# @load_models(
+#     (Order, {'id': 'line_item_id'}, 'line_item')
+#     )
+def cancel_line_item(line_item):
+    if not line_item.is_confirmed:
+        # only confirmed line items can be cancelled
+        abort(401)
```
Comment (jace) [inline boxoffice/views/order.py]: Isn't 401 "authorisation required" while 403 is "unauthorized". One of them presents a login prompt.

File: boxoffice/views/order.py [hunk 4]
```diff
@@ -314,6 +315,35 @@ def format_assignee(line_item):
...
+
+    if line_item.final_amount > Decimal('0'):
+        payment = OnlinePayment.query.filter_by(order=line_item.order).first()
```
Comment (jace) [inline boxoffice/views/order.py]: Use `.one()` since the result has to exist.

File: boxoffice/views/order.py [hunk 5]
```diff
@@ -314,6 +315,35 @@ def format_assignee(line_item):
...
+        rp_resp = razorpay.refund_payment(payment.pg_paymentid, line_item.final_amount)
+        if rp_resp.status_code == 200:
+            line_item.cancel()
+            cancel_transaction = PaymentTransaction(order=line_item.order, transaction_type=TRANSACTION_TYPE.REFUND,
+                online_payment=payment, amount=line_item.final_amount, currency=CURRENCY.INR)
+            db.session.add(line_item)
```
Comment (jace) [inline boxoffice/views/order.py]: Don't need to add since it's already in the session.

File: boxoffice/views/order.py [hunk 6]
```diff
@@ -314,6 +315,35 @@ def format_assignee(line_item):
...
+            db.session.add(cancel_transaction)
```
Comment (jace) [inline boxoffice/views/order.py]: This add isn't necessary either if any of the referred items have a backref to it.

File: boxoffice/views/order.py [hunk 7]
```diff
@@ -314,6 +315,35 @@ def format_assignee(line_item):
...
+            db.session.commit()
+        else:
+            raise APIError("Cancellation failed for order - {order}".format(order=line_item.order.id), 502)
+    else:
+        line_item.cancel()
+        db.session.add(line_item)
```
Comment (jace) [inline boxoffice/views/order.py]: As above.

File: boxoffice/mailclient.py [hunk 8]
```diff
@@ -30,3 +31,18 @@ def send_participant_assignment_mail(order_id, subject="Please tell us who's com
         msg.html = html
         msg.body = html2text(html)
         mail.send(msg)
+
+
+def send_line_item_cancellation_mail(line_item_id, subject="Ticket Cancellation"):
```
Comment (jace) [inline boxoffice/mailclient.py]: As this is a background job, use the `@job` decorator. The `app.test_request_context()` structure should only be used at an entry point, which this function is, but it's hard to tell without the decorator.

File: boxoffice/views/order.py [hunk 9]
```diff
@@ -314,6 +315,35 @@ def format_assignee(line_item):
...
+        db.session.commit()
+    boxofficeq.enqueue(send_line_item_cancellation_mail, line_item.id)
```
Comment (jace) [inline boxoffice/views/order.py]: Use `funcname.delay(…)` syntax as that is easier to read. Let the queue name come from the `@job` decorator.

File: boxoffice/views/order.py [hunk 10]
```diff
@@ -314,6 +315,35 @@ def format_assignee(line_item):
     return api_orders
 
 
+# TODO activate route when front-end becomes available
+# @app.route('/line_item/<line_item_id>/cancel', methods=['POST'])
+# @load_models(
+#     (Order, {'id': 'line_item_id'}, 'line_item')
+#     )
+def cancel_line_item(line_item):
+    if not line_item.is_confirmed:
```
Comment (shreyas-satish) [inline boxoffice/views/order.py]: Shouldn't we also disallow line items from being cancelled, after an event has completed (perhaps with a grace of one day)? If so, it makes sense to add `start_at` and `end_at` to an `ItemCollection`?

File: boxoffice/extapi/razorpay.py [hunk 11]
```diff
@@ -34,3 +34,13 @@ def capture_payment(paymentid, amount):
...
     return resp
+
+
+def refund_payment(paymentid, amount):
+    """
+    Sends a POST request to Razorpay, to initiate a refund
+    """
+    url = '{base_url}/{paymentid}/refund'.format(base_url=base_url, paymentid=paymentid)
+    # Razorpay requires the amount to be in paisa and of type integer
+    resp = requests.post(url, data={'amount': int(amount*100)}, auth=(app.config['RAZORPAY_KEY_ID'], app.config['RAZORPAY_KEY_SECRET']))
```
Comment (jace) [inline boxoffice/extapi/razorpay.py]: PyFlakes will ask that this be written as `amount * 100` (with spaces).

File: boxoffice/models/payment.py [hunk 12]
```diff
@@ -72,3 +72,7 @@ class PaymentTransaction(BaseMixin, db.Model):
 
 class CURRENCY(LabeledEnum):
     INR = (u"INR", __("INR"))
+
+
+class CURRENCY_SYMBOL(LabeledEnum):
+    INR = (u"INR", u'₹')
```
Comment (jace) [inline boxoffice/models/payment.py]: Double quotes for the rupee symbol as it is user facing.

File: boxoffice/templates/line_item_cancellation_mail.html [hunk 13]
```diff
@@ -0,0 +1,56 @@
...
+
+    <p>Hello {{ order.buyer_fullname }},</p>
+
+    <p>With reference to your order (Receipt No: {{ order.invoice_no }}), "{{ item_title }}" has been cancelled.</p>
+    {%- if is_paid %}
+      <p>A refund of {{currency_symbol}} {{line_item.final_amount}} has been initiated. You should receive the refund in 2-5 days.</p>
+    {%- endif %}
+  
+    <div class="terms">
+      <p class="terms-content">For further queries, please write to <a href="mailto:support@hasgeek.com">support@hasgeek.com</a></p>
```
Comment (jace) [inline boxoffice/templates/line_item_cancellation_mail.html]: Shouldn't this support address be in the org settings?

File: boxoffice/views/order.py [hunk 14]
```diff
@@ -314,6 +315,31 @@ def format_assignee(line_item):
...
+def cancel_line_item(line_item):
+    if not line_item.is_cancellable():
+        abort(403)
+
+    if line_item.final_amount > Decimal('0'):
+        payment = OnlinePayment.query.filter_by(order=line_item.order).one()
+        rp_resp = razorpay.refund_payment(payment.pg_paymentid, line_item.final_amount)
+        if rp_resp.status_code == 200:
+            line_item.cancel()
+            PaymentTransaction(order=line_item.order, transaction_type=TRANSACTION_TYPE.REFUND,
```
Comment (jace) [inline boxoffice/views/order.py]: Just for the sake of clarity, make this `db.session.add(PaymentTransaction(…))`

File: boxoffice/templates/line_item_cancellation_mail.html (hunk 13 above)
Comment (shreyas-satish) [inline boxoffice/templates/line_item_cancellation_mail.html]: Yes, reading it now from `org.contact_email`.

Comment (shreyas-satish) [inline boxoffice/templates/line_item_cancellation_mail.html]: Removed the paragraph, since it's easier for the user to just reply to that email.

**Pull request merged**